File: packages/quran-engine-js/src/topics.js
// @ts-check
/**
 * Two more ways to find your way around the Quran, both from the Quranic Universal Library.
 *
 * `QuranTopics` is 2,512 topics indexed **three independent ways**, and the independence is the
 * thing to understand before using it:
 *
 *  * `thematic` - the Clear Quran's thematic tree (Doctrine, Stories, The Unseen)
 *  * `ontology` - the Quranic Arabic Corpus ontology (Living Creation, Location, Event)
 *  * `index`    - a general A-Z index
 *
 * These are three trees over ONE pool of topics, not three partitions of it. A topic can be a
 * node in more than one (Adam sits in the thematic tree and in the ontology, 17 topics do), and
 * a tree's parent need not itself be listed in that tree (the A-Z index hangs "House of" under
 * Moses, a thematic topic). So every accessor that walks the hierarchy takes the tree you mean:
 * `parent(id, "ontology")` is a different question from `parent(id, "thematic")`, and both can
 * have an answer for the same topic. `families` says which indexes list a topic; passing no tree
 * uses the first of them, which is a convenience, not a fact about the data.
 *
 * `AyahThemes` is the other shape entirely: 1,049 **passages**, one short sentence describing a
 * run of ayahs ("Hypocrites and the consequences of hypocrisy", 2:8-16). Passages run in order
 * through a surah and do not nest, which is what makes them the right thing to print under an
 * ayah as "what is going on here", where a topic list would just be a pile of labels. They do
 * not tile the surah either: an ayah between two passages has none.
 *
 * This is a different corpus from `themes`, not a newer one. `themes` is 323 curated topics with
 * domains and categories; keep both, they answer differently.
 */

/**
 * @typedef {"thematic"|"ontology"|"index"} TopicTree
 */

/**
 * @typedef {Object} QulTopic
 * @property {number} id
 * @property {string} name
 * @property {string} arabic
 * @property {TopicTree[]} families                 the indexes listing this topic; 17 have two
 * @property {Record<TopicTree, number|null>} parents  one parent per tree, independently
 * @property {string} description
 * @property {string} wiki
 * @property {string[]} ayahs                       "surah:ayah" keys, in the corpus's order
 * @property {number[]} related
 */

/**
 * @typedef {Object} Passage
 * @property {number} surah
 * @property {number} from   first ayah
 * @property {number} to     last ayah
 * @property {string} theme  one sentence
 * @property {string} topic  the topic it sits under, "" when none
 */

export const TOPIC_FAMILIES = /** @type {const} */ (["thematic", "ontology", "index"]);

export class QuranTopics {
  /** @param {Object} [data] data/quran-topics.json @param {QulTopic[]} [data.topics] */
  constructor(data = {}) {
    this._topics = data.topics ?? [];
    /** @type {Map<number, QulTopic>} */
    this._byId = new Map(this._topics.map((t) => [t.id, t]));
    /** @type {Map<string, Map<number, number[]>>} */
    this._children = new Map();
    /** @type {Map<string, number[]>|null} */
    this._byAyah = null;
  }

  get isLoaded() {
    return this._topics.length > 0;
  }

  /** Every topic, in corpus order. @returns {QulTopic[]} */
  all() {
    return this._topics;
  }

  /** @param {number} id @returns {QulTopic|null} */
  topic(id) {
    return this._byId.get(id) ?? null;
  }

  /** The topics an index lists. @param {TopicTree} tree @returns {QulTopic[]} */
  inFamily(tree) {
    return this._topics.filter((t) => t.families.includes(tree));
  }

  /**
   * A tree's roots: topics it lists that have no parent in that same tree.
   * @param {TopicTree} tree @returns {QulTopic[]}
   */
  roots(tree) {
    return this._topics.filter((t) => t.families.includes(tree) && !t.parents[tree]);
  }

  /**
   * The parent in one tree. Defaults to the topic's first listed index, which is a convenience
   * for a caller that does not care; pass the tree explicitly when it matters.
   * @param {number} id @param {TopicTree} [tree] @returns {QulTopic|null}
   */
  parent(id, tree) {
    const topic = this.topic(id);
    if (!topic) return null;
    const which = tree ?? topic.families[0];
    const parentId = topic.parents[which];
    return parentId ? this.topic(parentId) : null;
  }

  /**
   * Direct children in one tree, in id order.
   * @param {number} id @param {TopicTree} [tree] @returns {QulTopic[]}
   */
  children(id, tree) {
    const which = tree ?? this.topic(id)?.families[0];
    if (!which) return [];
    return (this._childIndex(which).get(id) ?? []).map((cid) => /** @type QulTopic */ (this._byId.get(cid)));
  }

  /**
   * The chain from a topic up to its root in one tree, nearest first. Guards against a cycle in
   * the corpus rather than trusting it not to have one.
   * @param {number} id @param {TopicTree} [tree] @returns {QulTopic[]}
   */
  ancestors(id, tree) {
    const which = tree ?? this.topic(id)?.families[0];
    if (!which) return [];
    const out = [];
    const seen = new Set([id]);
    let current = this.parent(id, which);
    while (current && !seen.has(current.id)) {
      seen.add(current.id);
      out.push(current);
      current = this.parent(current.id, which);
    }
    return out;
  }

  /**
   * Every topic annotating this ayah, across all three indexes.
   * @param {number} surahId @param {number} ayahId @returns {QulTopic[]}
   */
  topicsFor(surahId, ayahId) {
    this._indexAyahs();
    const ids = this._byAyah?.get(`${surahId}:${ayahId}`) ?? [];
    return ids.map((id) => /** @type QulTopic */ (this._byId.get(id)));
  }

  /**
   * Name and Arabic-name substring search, exact-prefix hits first.
   * @param {string} query @param {number} [limit] @returns {QulTopic[]}
   */
  search(query, limit = 50) {
    const q = String(query ?? "").trim().toLowerCase();
    if (!q) return [];
    const starts = [];
    const contains = [];
    for (const topic of this._topics) {
      const name = topic.name.toLowerCase();
      if (name.startsWith(q)) starts.push(topic);
      else if (name.includes(q) || topic.arabic.includes(query)) contains.push(topic);
      if (starts.length >= limit) break;
    }
    return starts.concat(contains).slice(0, limit);
  }

  count() {
    const byFamily = { thematic: 0, ontology: 0, index: 0 };
    let refs = 0;
    for (const topic of this._topics) {
      for (const family of topic.families) byFamily[family] += 1;
      refs += topic.ayahs.length;
    }
    return { topics: this._topics.length, ...byFamily, references: refs };
  }

  /** @private Children of one tree, built on first use. */
  _childIndex(tree) {
    const cached = this._children.get(tree);
    if (cached) return cached;
    /** @type {Map<number, number[]>} */
    const map = new Map();
    for (const topic of this._topics) {
      const parentId = topic.parents[tree];
      if (!parentId) continue;
      const bucket = map.get(parentId);
      if (bucket) bucket.push(topic.id); else map.set(parentId, [topic.id]);
    }
    for (const bucket of map.values()) bucket.sort((a, b) => a - b);
    this._children.set(tree, map);
    return map;
  }

  /** @private */
  _indexAyahs() {
    if (this._byAyah) return;
    /** @type {Map<string, number[]>} */
    const map = new Map();
    for (const topic of this._topics) {
      for (const key of topic.ayahs) {
        const bucket = map.get(key);
        if (bucket) bucket.push(topic.id); else map.set(key, [topic.id]);
      }
    }
    this._byAyah = map;
  }
}

export class AyahThemes {
  /** @param {Object} [data] data/ayah-themes.json (surah -> passages) */
  constructor(data = {}) {
    this._data = data ?? {};
  }

  get isLoaded() {
    return Object.keys(this._data).length > 0;
  }

  /** A surah's passages, in order. @param {number} surahId @returns {Passage[]} */
  passages(surahId) {
    const rows = this._data[String(surahId)] ?? [];
    return rows.map((row) => ({ surah: surahId, from: row.from, to: row.to, theme: row.theme, topic: row.topic ?? "" }));
  }

  /**
   * The passage an ayah falls in, or null. Passages do not overlap, so this is the one answer.
   * @param {number} surahId @param {number} ayahId @returns {Passage|null}
   */
  passageFor(surahId, ayahId) {
    for (const passage of this.passages(surahId)) {
      if (ayahId >= passage.from && ayahId <= passage.to) return passage;
    }
    return null;
  }

  /** @param {string} query @param {number} [limit] @returns {Passage[]} */
  search(query, limit = 50) {
    const q = String(query ?? "").trim().toLowerCase();
    if (!q) return [];
    const out = [];
    for (const surah of Object.keys(this._data).map(Number).sort((a, b) => a - b)) {
      for (const passage of this.passages(surah)) {
        if (passage.theme.toLowerCase().includes(q) || passage.topic.toLowerCase().includes(q)) {
          out.push(passage);
          if (out.length >= limit) return out;
        }
      }
    }
    return out;
  }

  count() {
    let passages = 0;
    for (const rows of Object.values(this._data)) passages += rows.length;
    return { surahs: Object.keys(this._data).length, passages };
  }
}

File: packages/quran-engine-py/quran_engine/topics.py
"""QUL's topic indexes and the passage themes. Mirrors src/topics.js.

`QuranTopics` is 2,512 topics indexed THREE independent ways: the Clear Quran thematic tree, the
Quranic Arabic Corpus ontology, and a general A-Z index. Three trees over one pool of topics, not
three partitions of it: a topic can be a node in more than one (17 are), and a tree's parent need
not itself be listed in that tree. So every hierarchy accessor takes the tree you mean.

`AyahThemes` is 1,049 passages, one sentence per run of ayahs. They run in order and do not nest,
but they do not tile a surah either: an ayah between two passages has none.
"""
from __future__ import annotations
from typing import Optional

TOPIC_TREES = ("thematic", "ontology", "index")


class QulTopic(dict):
    """id, name, arabic, families, parents, description, wiki, ayahs, related."""


class Passage(dict):
    """surah, from_, to, theme, topic."""


class QuranTopics:
    def __init__(self, data: Optional[dict] = None):
        data = data or {}
        self._topics: list[QulTopic] = [QulTopic(t) for t in data.get("topics", [])]
        self._by_id = {t["id"]: t for t in self._topics}
        self._children: dict[str, dict[int, list[int]]] = {}
        self._by_ayah: Optional[dict[str, list[int]]] = None

    @property
    def is_loaded(self) -> bool:
        return bool(self._topics)

    def all(self) -> list[QulTopic]:
        return self._topics

    def topic(self, topic_id: int) -> Optional[QulTopic]:
        return self._by_id.get(topic_id)

    def in_family(self, tree: str) -> list[QulTopic]:
        """The topics an index lists."""
        return [t for t in self._topics if tree in t["families"]]

    def roots(self, tree: str) -> list[QulTopic]:
        """Topics an index lists that have no parent in that same tree."""
        return [t for t in self._topics if tree in t["families"] and not t["parents"].get(tree)]

    def parent(self, topic_id: int, tree: Optional[str] = None) -> Optional[QulTopic]:
        topic = self.topic(topic_id)
        if not topic:
            return None
        which = tree or (topic["families"][0] if topic["families"] else None)
        if not which:
            return None
        parent_id = topic["parents"].get(which)
        return self.topic(parent_id) if parent_id else None

    def children(self, topic_id: int, tree: Optional[str] = None) -> list[QulTopic]:
        topic = self.topic(topic_id)
        which = tree or (topic["families"][0] if topic and topic["families"] else None)
        if not which:
            return []
        return [self._by_id[c] for c in self._child_index(which).get(topic_id, [])]

    def ancestors(self, topic_id: int, tree: Optional[str] = None) -> list[QulTopic]:
        """The chain up to the root of one tree, nearest first; cycle-safe."""
        topic = self.topic(topic_id)
        which = tree or (topic["families"][0] if topic and topic["families"] else None)
        if not which:
            return []
        out: list[QulTopic] = []
        seen = {topic_id}
        current = self.parent(topic_id, which)
        while current and current["id"] not in seen:
            seen.add(current["id"])
            out.append(current)
            current = self.parent(current["id"], which)
        return out

    def topics_for(self, surah_id: int, ayah_id: int) -> list[QulTopic]:
        """Every topic annotating this ayah, across all three indexes."""
        self._index_ayahs()
        return [self._by_id[i] for i in (self._by_ayah or {}).get(f"{surah_id}:{ayah_id}", [])]

    def search(self, query: str, limit: int = 50) -> list[QulTopic]:
        q = (query or "").strip().lower()
        if not q:
            return []
        starts: list[QulTopic] = []
        contains: list[QulTopic] = []
        for topic in self._topics:
            name = topic["name"].lower()
            if name.startswith(q):
                starts.append(topic)
            elif q in name or query in topic["arabic"]:
                contains.append(topic)
            if len(starts) >= limit:
                break
        return (starts + contains)[:limit]

    def count(self) -> dict:
        by_family = {t: 0 for t in TOPIC_TREES}
        refs = 0
        for topic in self._topics:
            for family in topic["families"]:
                by_family[family] += 1
            refs += len(topic["ayahs"])
        return {"topics": len(self._topics), **by_family, "references": refs}

    # -- internals ---------------------------------------------------------------------

    def _child_index(self, tree: str) -> dict[int, list[int]]:
        cached = self._children.get(tree)
        if cached is not None:
            return cached
        table: dict[int, list[int]] = {}
        for topic in self._topics:
            parent_id = topic["parents"].get(tree)
            if parent_id:
                table.setdefault(parent_id, []).append(topic["id"])
        for bucket in table.values():
            bucket.sort()
        self._children[tree] = table
        return table

    def _index_ayahs(self) -> None:
        if self._by_ayah is not None:
            return
        table: dict[str, list[int]] = {}
        for topic in self._topics:
            for key in topic["ayahs"]:
                table.setdefault(key, []).append(topic["id"])
        self._by_ayah = table


class AyahThemes:
    def __init__(self, data: Optional[dict] = None):
        self._data = data or {}

    @property
    def is_loaded(self) -> bool:
        return bool(self._data)

    def passages(self, surah_id: int) -> list[Passage]:
        return [Passage({"surah": surah_id, "from_": row["from"], "to": row["to"],
                         "theme": row["theme"], "topic": row.get("topic", "")})
                for row in self._data.get(str(surah_id), [])]

    def passage_for(self, surah_id: int, ayah_id: int) -> Optional[Passage]:
        """Passages do not overlap, so this is the one answer."""
        for passage in self.passages(surah_id):
            if passage["from_"] <= ayah_id <= passage["to"]:
                return passage
        return None

    def search(self, query: str, limit: int = 50) -> list[Passage]:
        q = (query or "").strip().lower()
        if not q:
            return []
        out: list[Passage] = []
        for surah in sorted(int(k) for k in self._data):
            for passage in self.passages(surah):
                if q in passage["theme"].lower() or q in passage["topic"].lower():
                    out.append(passage)
                    if len(out) >= limit:
                        return out
        return out

    def count(self) -> dict:
        return {"surahs": len(self._data),
                "passages": sum(len(rows) for rows in self._data.values())}
